Ross-Robotics/rr_viz
# Bottom up Refactor

Hi so as I mentioned I'm start a PR for bottom up refactor. 

This PR includes the following:

- [Interactive_waypoints](https://github.com/JuliusSustarevas/interactive_waypoints) that rr_viz was based on was refactored and cleaned up very  well. These changes are not visible here because its a submodule. But the changes are as follows: save_to_msg, load_from_msg  are implemented thorughout. Movebase and saving/loading menu actions are now separated away from the list object.    These were important to keep waypoint click-menu and rr_viz button functionalities not clashing. Also  [ros_msgdict](https://github.com/JuliusSustarevas/ros_msgdict) was developed to aid saving/loading messages to file in a more consistent way. Key takeaway: **Waypoint is associated with a message type** For interactive_waypoints its Waypoint =PoseStamped  , WaypointList = PoseStampedArray.
- Using this new mentality interactive_waypoints  is still inherited and extended at rr_viz level (see RRTaskWaypoint etc.) Because we simply change message association from PoseStaped to rr_viz/TaskWaypoint , we can reuse saving/loading  mechanisms and in general only implement TaskWaypoint specific features e.g. set_tasksrv  etc.  This way we allow the backend of waypoints to be managed by interactive_waypoints, and task stuff, by RRTaskWaypoint
- Further tying in Actions on waypoints to rr_viz, RRTaskWaypoint/Lists are turned into QWidgets. See QWaypointWidget/List etc.  This allows model-view-control separation where QWaypointList only implements the qt-level logic. (Thus allowing more of it ina  clean way. e.g. state_checkers for buttons)
- rr_viz separted things a bit to make it more clear what each panel is for e.g. mission editing, commands etc. 
- Also theres some fancier markers
- As part of this the substories regarding **add here** button and **clean shutdown** were also addressed. 


To test this:
- You can check out this  by itself and run rr_viz_test.launch. Can kill rosmaster see how it reacts can kill mock nodes. 
You will need interactive_waypoints and ros_msgdict too. But ross_robo PR will have them. so can test there instead. 
- I am still working on how to test this with envirobot.
